### PR TITLE
Debug plugin

### PR DESCRIPTION
**Pull request by MichaelMraka**
Implementation of yum-debug-dump and yum-debug-restore.

### CODE REVIEW

File: plugins/debug.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+class DebugRestoreCommand(dnf.cli.Command):
+
+    aliases = ['debug-restore']
+    summary = _('restore packages recorded in debug-dump file')
+    usage = '[%s] [%s]' % (_('OPTIONS'), _('KEYWORDS'))
+
+    def run(self, args):
+        """Execute the command action here."""
+
+        parser = dnfpluginscore.ArgumentParser(self.aliases[0])
```
Comment (ignatenkobrain) [inline plugins/debug.py]: please make static function for parsing args as in other plugin: https://github.com/rpm-software-management/dnf-plugins-extras/blob/master/plugins/repomanage.py#L54

File: plugins/debug.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+        if opts.help_cmd:
+            print(parser.format_help())
+            return
+
+        if opts.filter_types:
+            opts.filter_types = set(
+                opts.filter_types.replace(",", " ").split())
+
+        self.base.fill_sack(load_system_repo=True,
```
Comment (ignatenkobrain) [inline plugins/debug.py]: use demands instead in `configure()`

File: plugins/debug.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        self.dump_rpmdb_versions(fobj)
+        fobj.close()
+
+        print(_("Output written to: %s") % filename)
+
+    @staticmethod
+    def dump_system_info(fobj):
+        fobj.write("%%%%SYSTEM INFO\n")
+        uname = os.uname()
+        rpm_ver = subprocess.check_output(["rpm", "--version"]).strip()
```
Comment (ignatenkobrain) [inline plugins/debug.py]: use `import rpm; rpm.__version__` instead

File: plugins/debug.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+# Red Hat, Inc.
+#
+
+from __future__ import absolute_import
+from __future__ import unicode_literals
+from dnfpluginscore import _, logger
+
+import dnf
+import dnf.cli
+import dnfpluginscore
```
Comment (ignatenkobrain) [inline plugins/debug.py]: dnfpluginsextras please ;)

File: plugins/debug.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+import dnfpluginscore.lib
```
Comment (ignatenkobrain) [inline plugins/debug.py]: unused import

File: plugins/debug.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+
+from __future__ import absolute_import
+from __future__ import unicode_literals
+from dnfpluginscore import _, logger
```
Comment (ignatenkobrain) [inline plugins/debug.py]: dnfpluginsextras please ;)

File: plugins/debug.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+class DebugDumpCommand(dnf.cli.Command):
+
+    aliases = ['debug-dump']
+    summary = _('dump information about installed rpm packages to file')
+    usage = '[%s] [%s]' % (_('OPTIONS'), _('KEYWORDS'))
+
+    def run(self, args):
+        """create debug txt file and compress it, if no filename specified
+           use dnf_debug_dump-<timestamp>.txt.gz by default"""
+        parser = dnfpluginscore.ArgumentParser(self.aliases[0])
```
Comment (ignatenkobrain) [inline plugins/debug.py]: dnfpluginsextras please ;)

File: plugins/debug.py (hunk 1 above)
Comment (ignatenkobrain) [inline plugins/debug.py]: dnfpluginsextras please ;)

Comment (MichaelMraka) [inline plugins/debug.py]: > please make static function for parsing args as in other plugin: https://github.com/rpm-software-management/dnf-plugins-extras/blob/master/plugins/repomanage.py#L54

This function is wrong it does not respond to `--help-cmd` correctly. The same bug is most likely in other plugins as well.

Comment (MichaelMraka) [inline plugins/debug.py]: >please make static function for parsing args

created _parse_args()

Comment (MichaelMraka) [inline plugins/debug.py]: > dnfpluginsextras please ;)

fixed

File: plugins/debug.py (hunk 2 above)
Comment (MichaelMraka) [inline plugins/debug.py]: > use demands instead in configure()

fixed

File: plugins/debug.py (hunk 3 above)
Comment (MichaelMraka) [inline plugins/debug.py]: >use import rpm; rpm.__version__ instead

fixed

File: plugins/debug.py (hunk 6 above)
Comment (MichaelMraka) [inline plugins/debug.py]: > dnfpluginsextras please ;)

fixed

File: plugins/debug.py (hunk 7 above)
Comment (MichaelMraka) [inline plugins/debug.py]: > dnfpluginsextras please ;)

fixed

File: plugins/debug.py (hunk 6 above)
Comment (ignatenkobrain) [inline plugins/debug.py]: not fixed ;)

Comment (MichaelMraka) [inline plugins/debug.py]: Oops, now fixed.

**Comment by ignatenkobrain** [comment]
YUM plugin:
```
%%%%RPMDB
  0:GConf2-3.2.6-11.fc22.x86_64
```

DNF Plugin:
```
%%%%RPMDB
  GConf2-0:3.2.6-11.fc22.x86_64
```

epoch not in the same place. probably in DNF plugin it's OK. what do you think?

**Comment by ignatenkobrain** [comment]
YUM:
```
%%%%RPMDB VERSIONS
  all: 3200:b50aa7cfcb5d762375647b0b4dd6232057a109fc
  yum: 15:07ef3ee9e97e04663c634bbb1e351806d2c77111
  yum-core: 3:3919f1aa6d322a115db055dfe92c46d337b5dee1
```

DNF:
```
%%%%RPMDB VERSIONS
  all: 3200:99bfaa692fded866976bafbd6a40613c329f27b2
```

I didn't have time to check what is it, but at least it's different.

**Comment by rholy** [comment]
BTW, if the dump files are not compatible with YUM, we should probably mention it in documentation at least.

**Comment by ignatenkobrain** [comment]
```
dnf-debug-dump version 1
```
There should be dnf-plugins-extras version I think

**Comment by ignatenkobrain** [comment]
> BTW, if the dump files are not compatible with YUM, we should probably mention it in documentation at least.
@rholy you are right

File: plugins/debug.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+        return
+
+
+class DebugRestoreCommand(dnf.cli.Command):
+
+    aliases = ['debug-restore']
+    summary = _('restore packages recorded in debug-dump file')
+    usage = '[%s] [%s]' % (_('OPTIONS'), _('KEYWORDS'))
+
+    def configure(self, args):
```
Comment (ignatenkobrain) [inline plugins/debug.py]: add demand for root, because we're installing/removing packages

**Comment by MichaelMraka** [comment]
> YUM:
> 0:GConf2-3.2.6-11.fc22.x86_64
>DNF Plugin:
> GConf2-0:3.2.6-11.fc22.x86_64

Yes, it's for purpose, epoch:name-... is not valid dnf package spec.

**Comment by MichaelMraka** [comment]
> BTW, if the dump files are not compatible with YUM, we should probably mention it in documentation at least.
(To prevent people loading YUM files and vice versa.)

They have different header so both yum and dnf refuse loading the other version.
But no problem add it to doc.

**Comment by rholy** [comment]
> They have different header so both yum and dnf refuse loading the other version.

Cool!

**Comment by MichaelMraka** [comment]
>>dnf-debug-dump version 1

>There should be dnf-plugins-extras version I think

No, it's debug file version not program version. The same version of file should be supported by different plugin versions.

**Comment by ignatenkobrain** [comment]
I think we need more general solution for help cmd, something like this:
```diff
diff --git a/plugins/dnfpluginsextras/__init__.py b/plugins/dnfpluginsextras/__init__.py
index e17855d..40af101 100644
--- a/plugins/dnfpluginsextras/__init__.py
+++ b/plugins/dnfpluginsextras/__init__.py
@@ -70,6 +70,8 @@ class ArgumentParser(argparse.ArgumentParser):
             opts = argparse.ArgumentParser.parse_args(self, args)
         except AttributeError as e:
             self.print_help()
+            if '--help-cmd' in args:
+                raise dnf.exceptions.Error
             raise dnf.exceptions.Error(str(e))
         return opts
```

**Comment by ignatenkobrain** [comment]
```python
$ dnf-3 debug-dump
Using metadata from Wed Feb 18 12:24:36 2015
Traceback (most recent call last):
  File "/bin/dnf-3", line 35, in <module>
    main.user_main(sys.argv[1:], exit_code=True)
  File "/usr/lib/python3.4/site-packages/dnf/cli/main.py", line 182, in user_main
    errcode = main(args)
  File "/usr/lib/python3.4/site-packages/dnf/cli/main.py", line 84, in main
    return _main(base, args)
  File "/usr/lib/python3.4/site-packages/dnf/cli/main.py", line 134, in _main
    cli.run()
  File "/usr/lib/python3.4/site-packages/dnf/cli/cli.py", line 1064, in run
    return self.command.run(self.base.extcmds)
  File "/usr/lib/python3.4/site-packages/dnf-plugins/debug.py", line 93, in run
    fobj.write(DEBUG_VERSION)
  File "/usr/lib64/python3.4/gzip.py", line 343, in write
    self.crc = zlib.crc32(data, self.crc) & 0xffffffff
TypeError: 'str' does not support the buffer interface
```

File: plugins/debug.py [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+    @staticmethod
+    def read_dump_file(filename):
+        if filename.endswith(".gz"):
+            fobj = gzip.GzipFile(filename)
+        else:
+            fobj = open(filename)
+
+        if fobj.readline() != DEBUG_VERSION:
+            logger.error(_("Bad dnf debug file: %s"), filename)
+            sys.exit(1)
```
Comment (ignatenkobrain) [inline plugins/debug.py]: use `raise dnf.exceptions.Error` instead

**Comment by MichaelMraka** [comment]
> TypeError: 'str' does not support the buffer interface

fixed

**Comment by ignatenkobrain** [comment]
Pushed to master. Thanks!

https://github.com/rpm-software-management/dnf-plugins-extras/commit/b0a2af298b028ea003d78fe0a2087b5becdf15ce

**Pull request closed without merging**